File: Endless/HSTSCache.m
/*
 * Endless
 * Copyright (c) 2014-2015 joshua stein <jcs@jcs.org>
 *
 * See LICENSE file for redistribution terms.
 */

#import "HSTSCache.h"
#import "NSString+IPAddress.h"

/* rfc6797 HTTP Strict Transport Security */

/* note that UIWebView has its own HSTS cache that comes preloaded with a big plist of hosts, but we can't change it or manually add to it */

// TODO: The preload list is stone-old and should be updated!
//
// A new one can be found here: https://chromium.googlesource.com/chromium/src/net/+/master/http/transport_security_state_static.json?format=TEXT
//
// And needs to be decoded:
// base64 --decode ~/Download/transport_security_state_static.json.txt
//
// And then filtered and transformed to plist format!
//
// TODO: This class should be rewritten in Swift with proper object serialization
// instead of a dictionary.

@implementation HSTSCache

static NSDictionary *_preloadedHosts;

+ (NSString *)hstsCachePath
{
	NSString *path = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) lastObject];
	return [path stringByAppendingPathComponent:@"hsts_cache.plist"];
}

- (HSTSCache *)init
{
	self = [super init];
	
	_dict = [[NSMutableDictionary alloc] init];
	
	return self;
}

+ (HSTSCache *)retrieve
{
	HSTSCache *hc = [[HSTSCache alloc] init];
	NSFileManager *fileManager = [NSFileManager defaultManager];
	if ([fileManager fileExistsAtPath:[[self class] hstsCachePath]]) {
		hc.dict = [NSMutableDictionary dictionaryWithContentsOfFile:[[self class] hstsCachePath]];
	}
	else {
		hc.dict = [[NSMutableDictionary alloc] initWithCapacity:50];
	}
	
	/* mix in preloaded */
	NSString *path = [[NSBundle mainBundle] pathForResource:@"hsts_preload" ofType:@"plist"];
	if ([[NSFileManager defaultManager] fileExistsAtPath:path]) {
		NSDictionary *tmp = [NSDictionary dictionaryWithContentsOfFile:path];
		for (NSString *host in [tmp allKeys]) {
			NSDictionary *hostdef = [tmp objectForKey:host];
			NSMutableDictionary *v = [[NSMutableDictionary alloc] init];
			
			[v setObject:[NSDate dateWithTimeIntervalSinceNow:(60 * 60 * 24 * 365)] forKey:HSTS_KEY_EXPIRATION];
			[v setObject:@YES forKey:HSTS_KEY_PRELOADED];
			
			NSNumber *is = [hostdef objectForKey:@"include_subdomains"];
			if ([is intValue] == 1) {
				[v setObject:@YES forKey:HSTS_KEY_ALLOW_SUBDOMAINS];
			}
			
			[[hc dict] setObject:v forKey:host];
		}
		
#ifdef TRACE_HSTS
		NSLog(@"[HSTSCache] locked and loaded with %lu preloaded hosts", [tmp count]);
#endif
	}
	else {
		NSLog(@"[HSTSCache] no preload plist at %@", path);
	}
	
	return hc;
}

- (void)persist
{
	@try {
		[self writeToFile:[[self class] hstsCachePath] atomically:YES];
	}
	@catch(NSException *e) {
		NSLog(@"[HSTSCache] failed persisting to file: %@", e);
	}
}

- (NSURL *)rewrittenURI:(NSURL *)URL
{
	if (![[URL scheme] isEqualToString:@"http"]) {
		return URL;
	}
	
	NSString *host = [[URL host] lowercaseString];
	NSString *matchHost = [host copy];
	
	/* 8.3: ignore when host is a bare ip address */
	if ([host isValidIPAddress]) {
		return URL;
	}
	
	NSDictionary *params = [self objectForKey:host];
	if (params == nil) {
		/* for a host of x.y.z.example.com, try y.z.example.com, z.example.com, example.com, etc. */
		NSArray *hostp = [host componentsSeparatedByString:@"."];
		for (int i = 1; i < [hostp count]; i++) {
			NSString *wc = [[hostp subarrayWithRange:NSMakeRange(i, [hostp count] - i)] componentsJoinedByString:@"."];
			
			if (((params = [self objectForKey:wc]) != nil) && [params objectForKey:HSTS_KEY_ALLOW_SUBDOMAINS]) {
				matchHost = wc;
				break;
			}
		}
	}
	
	if (params != nil) {
		NSDate *exp = [params objectForKey:HSTS_KEY_EXPIRATION];
		if ([exp timeIntervalSince1970] < [[NSDate date] timeIntervalSince1970]) {
#ifdef TRACE_HSTS
			NSLog(@"[HSTSCache] entry for %@ expired at %@", matchHost, exp);
#endif
			[self removeObjectForKey:matchHost];
			params = nil;
		}
	}
	
	if (params == nil) {
		return URL;
	}
	
	NSURLComponents *URLc = [NSURLComponents componentsWithURL:URL resolvingAgainstBaseURL:NO];
	
	[URLc setScheme:@"https"];
	
	/* 8.3.5: nullify port unless it's a non-standard one */
	if ([URLc port] != nil && [[URLc port] intValue] == 80) {
		[URLc setPort:nil];
	}
	
#ifdef TRACE_HSTS
	NSLog(@"[HSTSCache] %@rewrote %@ to %@", ([params objectForKey:HSTS_KEY_PRELOADED] ? @"[preloaded] " : @""), URL, [URLc URL]);
#endif
	
	return [URLc URL];
}

- (void)parseHSTSHeader:(NSString *)header forHost:(NSString *)host
{
	NSMutableDictionary *params = [[NSMutableDictionary alloc] initWithCapacity:3];
	host = [host lowercaseString];
	
	/* 8.1.1: reject caching when host is a bare ip address */
	if ([host isValidIPAddress])
		return;
	
#ifdef TRACE_HSTS
	NSLog(@"[HSTSCache] [%@] %@", host, header);
#endif

	NSArray *kvs = [header componentsSeparatedByString:@";"];
	for (NSString *kv in kvs) {
		NSArray *kvparts = [kv componentsSeparatedByString:@"="];
		NSString *key, *value;
		
		key = [kvparts[0] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
		
		if ([kvparts count] > 1) {
			value = [[kvparts[1] stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]] stringByReplacingOccurrencesOfString:@"\"" withString:@""];
		}
		
		if ([[key lowercaseString] isEqualToString:@"max-age"]) {
			long long age = [value longLongValue];

			if (age == 0) {
#ifdef TRACE_HSTS
				NSLog(@"[HSTSCache] [%@] got max-age=0, deleting", host);
#endif
				/* TODO: if a preloaded entry exists, cache a negative entry */
				[self removeObjectForKey:host];
				return;
			}
			else {
				NSDate *expire = [[NSDate date] dateByAddingTimeInterval:age];
				[params setObject:expire forKey:HSTS_KEY_EXPIRATION];
			}
		}
		else if ([[key lowercaseString] isEqualToString:@"includesubdomains"]) {
			[params setObject:@YES forKey:HSTS_KEY_ALLOW_SUBDOMAINS];
		}
		else if ([[key lowercaseString] isEqualToString:@"preload"] ||
			 [[key lowercaseString] isEqualToString:@""]) {
			/* ignore */
		}
		else {
#ifdef TRACE_HSTS
			NSLog(@"[HSTSCache] [%@] unknown parameter \"%@\"", host, key);
#endif
		}
	}
	
	if ([params objectForKey:HSTS_KEY_EXPIRATION]) {
		[self setValue:params forKey:host];
	}
}

/* NSMutableDictionary composition pass-throughs */

- (id)objectForKey:(id)aKey
{
	return [[self dict] objectForKey:aKey];
}

- (BOOL)writeToFile:(NSString *)path atomically:(BOOL)useAuxiliaryFile
{
	// FIXES BUGS #311 and #325:
	// Only write non-preloaded data, since we're ignoring it in #retrieve, anyway.
	// (OMFG! This took > 10 seconds to write and then got ignored!)
	NSMutableDictionary *temp = [NSMutableDictionary new];

	for (NSString *host in self.allKeys) {
		if (!self.dict[host][HSTS_KEY_PRELOADED]) {
			temp[host] = self.dict[host];
		}
	}

	@try {
		return [temp writeToFile:path atomically:useAuxiliaryFile];
	}
	@catch(NSException *e) {
		NSLog(@"[HSTSCache] failed persisting to file: %@", e);
	}
	
	return false;
}

- (void)setValue:(id)value forKey:(NSString *)key
{
	if (value != nil && key != nil)
		[[self dict] setValue:value forKey:key];
}

- (void)removeObjectForKey:(id)aKey
{
	[[self dict] removeObjectForKey:aKey];
}

- (NSArray *)allKeys
{
	return [[self dict] allKeys];
}

@end

File: OnionBrowser/HstsCache.swift
//
//  HstsCache.swift
//  OnionBrowser2
//
//  Created by Benjamin Erhart on 21.02.22.
//  Copyright © 2022 Tigas Ventures, LLC (Mike Tigas). All rights reserved.
//
//  This file is part of Onion Browser. See LICENSE file for redistribution terms.
//

import UIKit

@objc
class HstsCache: NSObject {

	@objc
	static let hstsHeader = "Strict-Transport-Security"

	@objc
	static let shared = HstsCache()


	private static let preloadedFile = Bundle.main.path(forResource: "hsts_preload", ofType: "plist")
	private static let cacheFile = FileManager.default.docsDir?.appendingPathComponent("hsts_cache.plist")


	private let preloaded: [String: Entry]
	private var cached: [String: Entry]


	private override init() {
		if let preloadedFile = Self.preloadedFile {
			preloaded = Self.load(from: URL(fileURLWithPath: preloadedFile))
		}
		else {
			preloaded = [:]
		}

		cached = Self.load(from: Self.cacheFile)

		print("[\(String(describing: type(of: self)))] Locked and loaded with \(preloaded.count) preloaded and \(cached.count) cached hosts.")

		super.init()
	}


	func persist() {
		guard let cacheFile = Self.cacheFile else {
			return
		}

		let encoder = PropertyListEncoder()
		encoder.outputFormat = .xml

		do {
			let data = try encoder.encode(cached)

			try data.write(to: cacheFile)

			print("[\(String(describing: type(of: self)))]#persisted successfully.")
		}
		catch {
			print("[\(String(describing: type(of: self)))]#persist error=\(error)")
		}
	}

	@objc(rewriteURL:)
	func rewrite(url: URL) -> URL {
		guard url.scheme == "http",
			  let host = url.host?.lowercased(),
			  !NSString(string: host).isValidIPAddress()
		else {
			return url
		}

		var matchHost = host

		var entry = entry(for: host)

		if entry == nil {
			// For a host of x.y.z.example.com, try y.z.example.com, z.example.com, example.com, etc.
			let hostp = host.components(separatedBy: ".")

			for i in 1 ..< hostp.count - 1 {
				let wc = hostp[i ..< hostp.count].joined(separator: ".")

				entry = self.entry(for: wc)

				if entry?.allowSubdomains ?? false {
					matchHost = wc
					break
				}
				else {
					entry = nil
				}
			}
		}

		guard let entry = entry else {
			return url
		}

		if entry.expiration < Date() {
			print("[\(String(describing: type(of: self)))] entry for \(matchHost) expired at \(entry.expiration)")

			remove(for: matchHost)

			return url
		}

		var urlc = URLComponents(url: url, resolvingAgainstBaseURL: false)
		urlc?.scheme = "https"

		// 8.3.5: nullify port unless it's a non-standard one.
		if urlc?.port == 80 {
			urlc?.port = nil
		}

		let newUrl = urlc?.url ?? url

		print("[\(String(describing: type(of: self)))] rewrote \(url) to \(newUrl)")

		return newUrl
	}

	@objc
	func parseHstsHeader(_ header: String, for host: String) {
		if NSString(string: host).isValidIPAddress() {
			return
		}

		let host = host.lowercased()

		print("[\(String(describing: type(of: self)))]#parseHstsHeader:for: [\(host)] \(header)")

		let kvs = header.components(separatedBy: ";")

		var age: Int64 = 0
		var allowSubdomains: Bool?

		for kv in kvs {
			let kvparts = kv.components(separatedBy: "=")

			let key = kvparts.first?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
			let value = kvparts.count < 2 ? nil : kvparts.last?.trimmingCharacters(in: .whitespacesAndNewlines) .replacingOccurrences(of: "\"", with: "")

			switch key {
			case "max-age":
				age = Int64(value ?? "0") ?? 0

			case "includesubdomains":
				allowSubdomains = true

			case "preload", "", nil:
				// Ignore.
				break

			default:
				print("[\(String(describing: type(of: self)))] [\(host)] unknown parameter \"\(key!)\"")
			}
		}

		if age <= 0 {
			print("[\(String(describing: type(of: self)))] [\(host)] got max-age=0, deleting")

			remove(for: host)

			return
		}

		cached[host] = Entry(expiration: Date().addingTimeInterval(Double(age)), allowSubdomains: allowSubdomains)
	}

	@objc
	func _testingOnlyEntry(_ host: String) -> [String: Any]? {
		guard let entry = entry(for: host) else {
			return nil
		}

		var result: [String: Any] = [
			"expiration": entry.expiration
		]

		if let allowSubdomains = entry.allowSubdomains {
			result["allowSubdomains"] = allowSubdomains
		}

		return result
	}


	// MARK: Private Methods

	private func entry(for host: String) -> Entry? {
		if let entry = cached[host] {
			return entry.ignore == true ? nil : entry
		}

		return preloaded[host]
	}

	private func remove(for host: String) {
		if preloaded[host] != nil {
			cached[host] = Entry(ignore: true)
			return
		}

		cached[host] = nil
	}

	private class func load(from url: URL?) -> [String: Entry] {
		if let url = url {
			do {
				return try PropertyListDecoder().decode([String: Entry].self, from: Data(contentsOf: url))
			}
			catch {
				print("[\(String(describing: type(of: self)))]#init error=\(error)")
			}
		}

		return [:]
	}
}

struct Entry: Codable {

	enum CodingKeys: String, CodingKey {
		case expiration
		case allowSubdomains
		case ignore
	}

	private var _expiration: Date?
	var expiration: Date {
		get {
			return _expiration ?? Date(timeIntervalSinceNow: 60 * 60 * 24 * 365)
		}
		set {
			_expiration = newValue
		}
	}

	var allowSubdomains: Bool?

	var ignore: Bool?

	init(expiration: Date? = nil, allowSubdomains: Bool? = nil, ignore: Bool? = nil) {
		_expiration = expiration
		self.allowSubdomains = allowSubdomains
		self.ignore = ignore
	}

	init(from decoder: Decoder) throws {
		let container = try? decoder.container(keyedBy: CodingKeys.self)

		_expiration = try? container?.decode(Date.self, forKey: .expiration)

		allowSubdomains = try? container?.decode(Bool.self, forKey: .allowSubdomains)

		ignore = try? container?.decode(Bool.self, forKey: .ignore)
	}

	func encode(to encoder: Encoder) throws {
		var container = encoder.container(keyedBy: CodingKeys.self)

		if let _expiration = _expiration {
			try container.encode(_expiration, forKey: .expiration)
		}

		if let allowSubdomains = allowSubdomains {
			try container.encode(allowSubdomains, forKey: .allowSubdomains)
		}

		if let ignore = ignore {
			try container.encode(ignore, forKey: .ignore)
		}
	}
}
